### ISSUE TITLE
After upgrading to Flutter 3.27.0, the usage of CachedNetworkImage causes error: SecurityError: Failed to execute 'texImage2D' on 'WebGL2RenderingContext': The image element contains cross-origin data, and may not be loaded.

### ISSUE THREAD

**Issue by rednikisfun** [Baseflow/flutter_cached_network_image#995]
## 🐛 Bug Report

I've upgraded my project's Flutter version to 3.27.0, and now, when I run the web version, I get a lot of errors like this:

`canvaskit.js:21 Uncaught (in promise) SecurityError: Failed to execute 'texImage2D' on 'WebGL2RenderingContext': The image element contains cross-origin data, and may not be loaded.`

UPD: Since in my project, CachedNetworkImage is heavily used, and there are quite a lot of images, too much `Failed to execute 'texImage2D' on 'WebGL2RenderingContext` causes `Additional error: RuntimeError: memory access out of bounds` and overall crash of the app.

### Expected behavior
It should load an image normally, like it used to in the previous Flutter versions

### Reproduction steps
1. Create a project with flutter version 3.27.0
2. Use CachedNetworkImage
3. Run project in web

### Configuration

**Version:** 3.4.1

**Platform:** 
- [ ] :web

**Comment by rednikisfun** [comment]
Changing from:

`CachedNetworkImage(
                  imageUrl:
                      slides[itemIndex].image.desktop.fromStringLang(lang),
                  fit: BoxFit.cover,
                  alignment: Alignment.center,
                ),`

to 

`CachedNetworkImage(
                  imageUrl:
                      slides[itemIndex].image.desktop.fromStringLang(lang),
                  imageRenderMethodForWeb: ImageRenderMethodForWeb.HttpGet, // <-- Add this line
                  fit: BoxFit.cover,
                  alignment: Alignment.center,
                ),`
                
seems to solve the problem provisionally

**Comment by lordarcadius** [comment]
> Changing from:
> 
> `CachedNetworkImage( imageUrl: slides[itemIndex].image.desktop.fromStringLang(lang), fit: BoxFit.cover, alignment: Alignment.center, ),`
> 
> to
> 
> `CachedNetworkImage( imageUrl: slides[itemIndex].image.desktop.fromStringLang(lang), imageRenderMethodForWeb: ImageRenderMethodForWeb.HttpGet, // <-- Add this line fit: BoxFit.cover, alignment: Alignment.center, ),`
> 
> seems to solve the problem provisionally

I am not able to import "ImageRenderMethodForWeb". It says undefined name. Can you share the import path?

**Comment by rednikisfun** [comment]
@lordarcadius 
> I am not able to import "ImageRenderMethodForWeb". It says undefined name. Can you share the import path?


Right, forgot to mention, you need to use `import 'package:cached_network_image_platform_interface/cached_network_image_platform_interface.dart';` for it to have access to `ImageRenderMethodForWeb`.

I've also forked the `cached_network_image` package, and set the `ImageRenderMethodForWeb.HttpGet` as default. If your project heavily relies on `cached_network_image`, I would recommend to temporarily use this [fork](https://github.com/Neotech-iHoops/flutter_cached_network_image.git) until `flutter_cached_network_image` team roll out a fix.

I've also created a [PR](https://github.com/Baseflow/flutter_cached_network_image/pull/996), if they accept it, the `ImageRenderMethodForWeb.HttpGet` will be set by default.

**Comment by eduardofcr** [comment]
same problem after upgrade flutter to 3.27.0

**Comment by zedaraf** [comment]
> @lordarcadius
> 
> > I am not able to import "ImageRenderMethodForWeb". It says undefined name. Can you share the import path?
> 
> Right, forgot to mention, you need to use `import 'package:cached_network_image_platform_interface/cached_network_image_platform_interface.dart';` for it to have access to `ImageRenderMethodForWeb`.
> 
> I've also forked the `cached_network_image` package, and set the `ImageRenderMethodForWeb.HttpGet` as default. If your project heavily relies on `cached_network_image`, I would recommend to temporarily use this [fork](https://github.com/Neotech-iHoops/flutter_cached_network_image.git) until `flutter_cached_network_image` team roll out a fix.
> 
> I've also created a [PR](https://github.com/Baseflow/flutter_cached_network_image/pull/996), if they accept it, the `ImageRenderMethodForWeb.HttpGet` will be set by default.

Hey; how did you add this fork to your pubspec.yaml exactly? @lordarcadius

**Comment by rednikisfun** [comment]
@ZiyadF296 


  cached_network_image:
    git:
      url: https://github.com/Neotech-iHoops/flutter_cached_network_image.git
      ref: develop
      path: cached_network_image

**Comment by djol** [comment]
While @rednikisfun's fix works, it ? seems to cause a new error under Safari (v18.2) (works fine with Chrome). When an image isn't found, the errorWidget does not seem to be used / in my case the placeholder widget (I'm using a CircularProgressIndicator()) instead persists. Prior to the 3.27.0 bug, this worked fine across all browsers.

<img width="323" alt="Screenshot 2024-12-19 at 12 39 41 pm" src="https://github.com/user-attachments/assets/25aa8da0-a9ce-4a28-89e0-741b535ac74c" />

**Comment by zedaraf** [comment]
I think there is a hotfix on the way directly to stable for this issue.

**Comment by martin-bautista-nimble** [comment]
Do we have an ETA For that fix? @ZiyadF296

**Comment by zedaraf** [comment]
@martin-bautista-nimble 

It was mentioned [here](https://github.com/flutter/flutter/issues/160127#issuecomment-2547012403).

Flutter PR [#160127](https://github.com/flutter/flutter/issues/160127) should solve it. You can track this PR until it reaches your branch, but it'll probably be pushed directly to stable.

**Comment by tfozo** [comment]
so i created a new file called cross_origin.dart in it

```import 'dart:async';
import 'dart:typed_data';
import 'dart:ui' as ui;
import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';
import 'package:web/web.dart' as web;

class CrossOriginNetworkImage extends ImageProvider<CrossOriginNetworkImage> {
  final String url;

  CrossOriginNetworkImage(this.url);

  @override
  Future<CrossOriginNetworkImage> obtainKey(ImageConfiguration configuration) {
    return SynchronousFuture(this);
  }

  @override
  ImageStreamCompleter loadImage(
      CrossOriginNetworkImage key, ImageDecoderCallback decode) {
    final completer = Completer<ImageInfo>();

    final imageElement = web.HTMLImageElement()
      ..src = url
      ..crossOrigin = 'anonymous';

    imageElement.onLoad.listen((_) async {
      final canvas = web.HTMLCanvasElement();
      canvas.width = imageElement.width;
      canvas.height = imageElement.height;

      final ctx = canvas.context2D;
      ctx.drawImage(imageElement, 0, 0);

      // Convert canvas to a data URL and extract bytes.
      final dataUrl = canvas.toDataUrl('image/png');

      final bytes = UriData.parse(dataUrl).contentAsBytes();

      // Decode image bytes into a Flutter ui.Image.
      final codec = await ui.instantiateImageCodec(Uint8List.fromList(bytes));
      final frame = await codec.getNextFrame();

      completer.complete(ImageInfo(image: frame.image, scale: 1.0));
    });

    imageElement.onError.listen((_) {
      completer.completeError(Exception("Failed to load image: $url"));
    });

    return OneFrameImageStreamCompleter(completer.future);
  }
}
```

and then I **replaced all my images** that were using CachedImageNetworks with

```
CachedNetworkImage(
                                        imageUrl: _images[index].imageUrl!,
                                        fit: BoxFit.cover,
                                        imageBuilder: (context, imageProvider) {
                                          return Image(
                                            image: CrossOriginNetworkImage(
                                                _images[index].imageUrl!),
                                            fit: BoxFit.cover,
                                          );
                                        },
                                        placeholder: (context, url) => Center(
                                          child: SpinKitPulse(
                                            color: Colors.amber[100],
                                            size: 50.0,
                                          ),
                                        ),
                                        errorWidget: (context, url, error) =>
                                            const Icon(Icons.error),
                                      )
```

the only change i made is the imageBuilder attribute

it should fix it but I honestly don't know how it fixed it.
I'm waiting for a fix from CachedNetworkImage but until that is implemented i just use this to work on my other features.

**Comment by AndreiMisiukevich** [comment]
@tfozo, the fix is in Flutter, and it should be included in the next stable release. I saw a cherry-pick.

**Comment by tfozo** [comment]
@AndreiMisiukevich now what do I do? 

Edit: Ahh, I'll keep an eye out for the update. Thanks again!

**Comment by AndreiMisiukevich** [comment]
> @AndreiMisiukevich now what do I do?
> 
> Edit: Ahh, I'll keep an eye out for the update. Thanks again!

I'd propose to stick to the previous stable version until it's resolved  - as we did 👍

**Comment by nekomaruh** [comment]
It's a weird behavior, my web page broke in production without even notice. Meanwhile had to switch to Image.network.

**Comment by vidhi002** [comment]
same issue

**Comment by AhabLives** [comment]
The issue is related to the CanvasKit Web Renderer (3.27.1)
"flutter run -d chrome --web-renderer html" will  fix the problem locally.
I have not tested build/deploy with "html".

**Comment by rafalplonka** [comment]
The html renderer will be removed though 

https://github.com/flutter/flutter/issues/145954#issuecomment-2551953495

**Comment by nikb7** [comment]
Flutter 3.27.2 has fixed this issue.

**Comment by rednikisfun** [comment]
I've upgraded to 3.27.2, and the latest `cached_network_image: ^3.4.1` package is working without issues, therefore, I am closing the issue.
